src/sys.rs

```diff
@@ -0,0 +1,374 @@
+    }
+
+    /// Register a new file descriptor for polling
+    ///
+    /// The file descriptor will be registered with given interest,
+    /// mode and token. This function will fail if given a
+    /// bad file descriptor or if the provided file descriptor is already
+    /// registered.
+    ///
+    /// # Leaking tokens
+    ///
+    /// If your event source is dropped without being unregistered, the token
+    /// passed in here will remain on the heap and continue to be used by the
+    /// polling system even though no event source will match it.
+    pub fn register(
+        &self,
+        #[cfg(unix)] fd: impl AsFd,
+        #[cfg(windows)] fd: impl AsSocket,
+        interest: Interest,
+        mode: Mode,
+        token: Token,
+    ) -> crate::Result<()> {
+        let raw = {
+            #[cfg(unix)]
+            {
+                fd.as_fd().as_raw_fd()
+            }
+
+            #[cfg(windows)]
+            {
+                fd.as_socket().as_raw_socket()
+            }
+        };
+
+        self.poller
+            .add_with_mode(raw, cvt_interest(interest, token), cvt_mode(mode))?;
+
+        Ok(())
+    }
+
+    /// Update the registration for a file descriptor
+    ///
+    /// This allows you to change the interest, mode or token of a file
+    /// descriptor. Fails if the provided fd is not currently registered.
+    ///
+    /// See note on [`register()`](Self::register()) regarding leaking.
+    pub fn reregister(
+        &self,
+        #[cfg(unix)] fd: impl AsFd,
+        #[cfg(windows)] fd: impl AsSocket,
+        interest: Interest,
+        mode: Mode,
+        token: Token,
+    ) -> crate::Result<()> {
+        let raw = {
+            #[cfg(unix)]
+            {
+                fd.as_fd().as_raw_fd()
+            }
+
+            #[cfg(windows)]
+            {
+                fd.as_socket().as_raw_socket()
+            }
+        };
+
+        self.poller
+            .modify_with_mode(raw, cvt_interest(interest, token), cvt_mode(mode))?;
+
+        Ok(())
+    }
+
+    /// Unregister a file descriptor
+    ///
+    /// This file descriptor will no longer generate events. Fails if the
+    /// provided file descriptor is not currently registered.
+    pub fn unregister(
+        &self,
+        #[cfg(unix)] fd: impl AsFd,
+        #[cfg(windows)] fd: impl AsSocket,
+    ) -> crate::Result<()> {
+        let raw = {
+            #[cfg(unix)]
+            {
+                fd.as_fd().as_raw_fd()
+            }
+
+            #[cfg(windows)]
+            {
+                fd.as_socket().as_raw_socket()
+            }
+        };
+        self.poller.delete(raw)?;
+
+        Ok(())
+    }
+
+    /// Get a thread-safe handle which can be used to wake up the `Poll`.
+    pub(crate) fn notifier(&self) -> Notifier {
+        Notifier(self.poller.clone())
+    }
+}
+
+/// Thread-safe handle which can be used to wake up the `Poll`.
+#[derive(Clone)]
+pub(crate) struct Notifier(Arc<Poller>);
```

Comment:
It's probably possible to rework the `Ping` event source to use that, by integrating it more tightly like the timers.

Though that can probably be done in a later PR.